File: android/app/src/main/java/dev/scorpion7slayer/modelsmeter/CelebrationDetector.java
package dev.scorpion7slayer.modelsmeter;

/** Pure decision logic for unexpected allowance refills and reset-credit increases. */
public final class CelebrationDetector {
    public static final int FIVE_HOUR = 1;
    public static final int WEEKLY = 1 << 1;
    public static final int MONTHLY = 1 << 2;

    private CelebrationDetector() {
    }

    /**
     * Returns a bit mask for windows that became completely available before their previously
     * advertised reset deadline. A missing deadline is deliberately ignored because a natural
     * reset cannot be distinguished from an external refill in that case.
     */
    public static int detectUnexpectedRefills(UsageSnapshot previous, UsageSnapshot current) {
        if (previous == null || current == null || current.fetchedAtMillis <= 0L) return 0;
        int refills = 0;
        if (isUnexpectedRefill(previous, previous.fiveHour, current.fiveHour,
                current.fetchedAtMillis)) {
            refills |= FIVE_HOUR;
        }
        if (isUnexpectedRefill(previous, previous.weekly, current.weekly,
                current.fetchedAtMillis)) {
            refills |= WEEKLY;
        }
        if (isUnexpectedRefill(previous, previous.monthly, current.monthly,
                current.fetchedAtMillis)) {
            refills |= MONTHLY;
        }
        return refills;
    }

    public static int resetCreditsAdded(int previous, int current) {
        if (previous < 0 || current <= previous) return 0;
        return current - previous;
    }

    public static int withoutUserResetRefills(int refills, long observedAtMillis,
            long fiveHourSuppressUntil, long weeklySuppressUntil) {
        return withoutUserResetRefills(refills, observedAtMillis, fiveHourSuppressUntil,
                weeklySuppressUntil, 0L);
    }

    public static int withoutUserResetRefills(int refills, long observedAtMillis,
            long fiveHourSuppressUntil, long weeklySuppressUntil, long monthlySuppressUntil) {
        if (fiveHourSuppressUntil > observedAtMillis) {
            refills &= ~FIVE_HOUR;
        }
        if (weeklySuppressUntil > observedAtMillis) {
            refills &= ~WEEKLY;
        }
        if (monthlySuppressUntil > observedAtMillis) {
            refills &= ~MONTHLY;
        }
        return refills;
    }

    static long expectedResetMillis(UsageSnapshot snapshot, UsageWindow window) {
        if (snapshot == null || window == null) return 0L;
        long resetAt = window.resetAtMillis();
        if (resetAt > 0L) return resetAt;
        if (snapshot.fetchedAtMillis <= 0L || window.resetAfterSeconds <= 0L) return 0L;
        if (window.resetAfterSeconds > (Long.MAX_VALUE - snapshot.fetchedAtMillis) / 1000L) {
            return Long.MAX_VALUE;
        }
        return snapshot.fetchedAtMillis + window.resetAfterSeconds * 1000L;
    }

    private static boolean isUnexpectedRefill(UsageSnapshot previous, UsageWindow oldWindow,
            UsageWindow newWindow, long observedAtMillis) {
        if (oldWindow == null || newWindow == null
                || oldWindow.usedPercent <= 0 || newWindow.usedPercent != 0) {
            return false;
        }
        long expectedReset = expectedResetMillis(previous, oldWindow);
        return expectedReset > observedAtMillis;
    }
}

File: ios/ModelsMeter/Infrastructure/CelebrationDetector.swift
import ModelsMeterCore
import Foundation

/// Pure decision logic for unexpected allowance refills and reset-credit increases
/// (parity with upstream CelebrationDetector from Models Meter 2.1+).
nonisolated public enum CelebrationDetector {
    public struct RefillMask: OptionSet, Sendable, Equatable {
        public let rawValue: Int

        public init(rawValue: Int) {
            self.rawValue = rawValue
        }

        public static let fiveHour = RefillMask(rawValue: 1 << 0)
        public static let weekly = RefillMask(rawValue: 1 << 1)
        public static let monthly = RefillMask(rawValue: 1 << 2)
    }

    public static let unknownUserResetSuppression: TimeInterval = 15 * 60

    /// Windows that became completely available before their previously advertised reset deadline.
    public static func detectUnexpectedRefills(
        previous: UsageSnapshot?,
        current: UsageSnapshot
    ) -> RefillMask {
        guard let previous else { return [] }
        var mask: RefillMask = []
        if isUnexpectedRefill(
            previous: previous,
            oldWindow: previous.fiveHour,
            newWindow: current.fiveHour,
            observedAt: current.fetchedAt
        ) {
            mask.insert(.fiveHour)
        }
        if isUnexpectedRefill(
            previous: previous,
            oldWindow: previous.weekly,
            newWindow: current.weekly,
            observedAt: current.fetchedAt
        ) {
            mask.insert(.weekly)
        }
        if isUnexpectedRefill(
            previous: previous,
            oldWindow: previous.monthly,
            newWindow: current.monthly,
            observedAt: current.fetchedAt
        ) {
            mask.insert(.monthly)
        }
        return mask
    }

    public static func resetCreditsAdded(previous: Int, current: Int) -> Int {
        guard previous >= 0, current > previous else { return 0 }
        return current - previous
    }

    public static func withoutUserResetRefills(
        _ refills: RefillMask,
        observedAt: Date,
        fiveHourSuppressUntil: Date?,
        weeklySuppressUntil: Date?,
        monthlySuppressUntil: Date? = nil
    ) -> RefillMask {
        var result = refills
        if let fiveHourSuppressUntil, fiveHourSuppressUntil > observedAt {
            result.remove(.fiveHour)
        }
        if let weeklySuppressUntil, weeklySuppressUntil > observedAt {
            result.remove(.weekly)
        }
        if let monthlySuppressUntil, monthlySuppressUntil > observedAt {
            result.remove(.monthly)
        }
        return result
    }

    public static func expectedResetDate(
        snapshot: UsageSnapshot,
        window: UsageWindow?
    ) -> Date? {
        guard let window else { return nil }
        return window.effectiveResetDate(relativeTo: snapshot.fetchedAt)
    }

    public static func userResetSuppressionDeadline(
        snapshot: UsageSnapshot,
        window: UsageWindow?,
        now: Date
    ) -> Date? {
        guard let window, window.usedPercent > 0 else { return nil }
        if let expected = expectedResetDate(snapshot: snapshot, window: window), expected > now {
            return expected
        }
        return now.addingTimeInterval(unknownUserResetSuppression)
    }

    private static func isUnexpectedRefill(
        previous: UsageSnapshot,
        oldWindow: UsageWindow?,
        newWindow: UsageWindow?,
        observedAt: Date
    ) -> Bool {
        guard let oldWindow, let newWindow else { return false }
        guard oldWindow.usedPercent > 0, newWindow.usedPercent == 0 else { return false }
        guard let expectedReset = expectedResetDate(snapshot: previous, window: oldWindow) else {
            return false
        }
        return expectedReset > observedAt
    }
}
